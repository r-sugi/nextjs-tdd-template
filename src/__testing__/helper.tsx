export function toMock<T extends jest.Func>(
  fn: T
): jest.Mock<ReturnType<T>, Parameters<T>> {
  return fn as unknown as jest.Mock<ReturnType<T>, Parameters<T>>;
}

export function toSpyWithMock<T extends (...args: any[]) => any>(
  obj: any,
  method: string,
  mockImpl: T
): jest.SpyInstance {
  const spy = jest.spyOn(obj, method);
  spy.mockImplementation(mockImpl);
  return spy;
}
