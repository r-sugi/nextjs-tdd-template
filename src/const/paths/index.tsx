const title = (page: string | undefined) => {
  const site = "myWebSite | マイウェブサイト";
  return page ? `${page} | ${site}` : site;
};

const publicPages = {
  index: {
    path: () => "/",
    title: () => title("トップページ"),
    description: () => "トップページです。",
  },
};

const loginRequiredPages = {
  mypage: {
    path: () => "/mypage",
    title: () => title("マイページ"),
    description: () => "マイページ",
  },
  mypageResignMemberInput: {
    path: () => "/mypage/resign-member/input",
    title: () => title("退会申請入力ページ"),
    description: () => "退会申請入力ページ",
  },
  mypageResignMemberConfirm: {
    path: () => "/mypage/resign-member/confirm",
    title: () => title("退会申請確認ページ"),
    description: () => "退会申請確認ページ",
  },
};

const adminSecretPages = {
  members: {
    path: () => "/admin/members",
    title: () => title("管理画面 会員一覧"),
    description: () => "管理画面 会員一覧",
  },
};

export { publicPages, loginRequiredPages, adminSecretPages };
