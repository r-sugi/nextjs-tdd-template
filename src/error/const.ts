type RuntimeError = {
	cause?: unknown | string;
	stack?: string;
};

export type AppErrorMessage = {
	title?: string;
	message?: string;
	myErrorCode: string;
	level: "fatal" | "error" | "warning" | "info" | "debug";
	// runtime error
	runtime?: RuntimeError;
};

// TODO: 下記の型だとEE99までサジェストされないので、直したい
type AppErrorObject = Record<string, AppErrorMessage>;

type AppError = {
	SYSTEM: {
		RECOVERABLE: AppErrorObject;
		UNRECOVERABLE: AppErrorObject;
	};
	BUSINESS: {
		RECOVERABLE: AppErrorObject;
		UNRECOVERABLE: AppErrorObject;
	};
};

export const APP_ERROR: AppError = {
	SYSTEM: {
		RECOVERABLE: {},
		UNRECOVERABLE: {
			EE98: {
				// 非同期処理が完了しなかったエラー
				title: "エラーが発生しました",
				message:
					"しばらく時間をおいてから再度お試しください。解消しない場合は開発者にお問い合わせください",
				myErrorCode: "EE98",
				level: "fatal",
			},
			EE99: {
				// 予期せぬエラー
				title: "エラーが発生しました",
				message:
					"しばらく時間をおいてから再度お試しください。解消しない場合は開発者にお問い合わせください",
				myErrorCode: "EE99",
				level: "fatal",
			},
		},
	},
	BUSINESS: {
		RECOVERABLE: {
			AU10: {
				// firebaseエラー
				title: "サインアップエラー",
				message: "メールアドレスはすでに利用されています。",
				myErrorCode: "AU10",
				level: "error",
			},
			AU20: {
				// firebaseエラー
				title: "ログインエラー",
				message: "メールアドレスまたはパスワードが間違っています。",
				myErrorCode: "AU20",
				level: "error",
			},
		},
		UNRECOVERABLE: {
			AU99: {
				// firebaseエラー
				title: "ネットワークエラー",
				message:
					"認証処理でネットワークエラーが発生しました。時間をおいて再度試してください",
				myErrorCode: "AU99",
				level: "fatal",
			},
		},
	},
};
