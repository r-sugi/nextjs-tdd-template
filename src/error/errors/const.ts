export const ERROR_CODE = {
	INTERNAL_SERVER_ERROR: "INTERNAL_SERVER_ERROR",
} as const;
// type ErrorCode = (typeof ERROR_CODE)[keyof typeof ERROR_CODE];

type AppErrorMessage = {
	title?: string;
	message?: string;
	myErrorCode: string;
};

export const MY_ERROR: Readonly<Record<string, AppErrorMessage>> = {
	EER99: {
		title: "エラーが発生しました",
		message:
			"しばらく時間をおいてから再度お試しください。解消しない場合は開発者にお問い合わせください",
		myErrorCode: "EER99",
	},
};
