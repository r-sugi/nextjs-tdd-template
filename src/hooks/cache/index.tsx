import { SessionStorageKeyValues } from "./type";
import { selectWebStorage, WebStorageType } from "./useWebStorage";

export const useGetCache = <
  K extends keyof SessionStorageKeyValues,
  V extends SessionStorageKeyValues[K]
>(
  key: K,
  storageType?: WebStorageType
): V => {
  const storage = selectWebStorage(storageType);
  const parsedValue = () => {
    const value = storage.getItem(key) as string | null;
    if (value !== null) {
      return JSON.parse(value) as V;
    }
    return value;
  };

  const object = parsedValue();
  if (object === null) {
    throw new Error("Failed to parse storage value");
  }
  return object;
};

export const useSetCache = <
  K extends keyof SessionStorageKeyValues,
  S extends SessionStorageKeyValues[K]
>(
  key: K,
  storageType?: WebStorageType
): ((value: S) => void) => {
  const storage = selectWebStorage(storageType);

  const setValue = (object: S) => {
    try {
      storage.setItem(key, JSON.stringify(object));
    } catch (error) {
      throw new Error("Failed to set storage value");
    }
  };
  return setValue;
};
