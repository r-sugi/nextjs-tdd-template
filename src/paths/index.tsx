const title = (page: string | undefined) => {
  const site = "myWebSite | マイウェブサイト";
  return page ? `${page} | ${site}` : site;
};

const publicPages = {
  index: {
    path: () => "/",
    title: () => title("トップページ"),
    description: () => "トップページです。",
  },
  posts: {
    path: () => "/posts",
    title: () => title("投稿一覧"),
    description: () => "投稿一覧のページです。",
  },
  postId: {
    path: (id: string | number) => `/posts/${id}`,
    title: () =>
      title(`投稿詳細`),
    description: () => "投稿詳細のページです。",
  },
};

export { publicPages };
