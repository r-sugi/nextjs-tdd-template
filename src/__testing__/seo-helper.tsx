import type { ReactElement } from "react";

import { toSpyWithMock } from "./helper";

export function mockNextHead() {
	toSpyWithMock(
		require("next/head"),
		"default",
		({ children }: { children: Array<ReactElement> }) => {
			return <>{children}</>;
		},
	);
}

export function assertSeoTags({
	titleText,
	descriptionText,
	ogUrlText,
}: {
	titleText: string;
	descriptionText: string;
	ogUrlText: string;
}) {
	const titleTag = document.querySelector("title");
	expect(titleTag).toBeInTheDocument();
	expect(titleTag?.textContent).toBe(titleText);

	// description
	const metaDescription = document.querySelector("meta[name='description']");
	expect(metaDescription).toBeInTheDocument();
	expect(metaDescription?.attributes?.getNamedItem("content")?.value).toBe(
		descriptionText,
	);

	// og:url
	const ogUrl = document.querySelector("meta[property='og:url']");
	expect(ogUrl).toBeInTheDocument();
	expect(ogUrl?.attributes?.getNamedItem("content")?.value).toBe(ogUrlText);
}
