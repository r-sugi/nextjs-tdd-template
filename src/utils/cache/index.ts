import { NoCacheError } from "./error";
import { SessionStorageKeyValues } from "./type";
import { selectWebStorage, WebStorageType } from "./useWebStorage";

export const getCache = <
  K extends keyof SessionStorageKeyValues,
  V extends SessionStorageKeyValues[K]
>(
  key: K,
  storageType?: WebStorageType
): V => {
  const storage = selectWebStorage(storageType);
  const parsedValue = () => {
    const value = storage.getItem(key) as string | null;
    if (value !== null) {
      return JSON.parse(value) as V;
    }
    return value;
  };

  const object = parsedValue();
  if (object === null) {
    throw new NoCacheError("Failed to parse storage value");
  }
  return object;
};

export const setCache = <
  K extends keyof SessionStorageKeyValues,
  S extends SessionStorageKeyValues[K]
>(
  key: K,
  object: S,
  storageType?: WebStorageType
): void => {
  const storage = selectWebStorage(storageType);

  try {
    storage.setItem(key, JSON.stringify(object));
  } catch (error) {
    throw new NoCacheError("Failed to set storage value");
  }
};
