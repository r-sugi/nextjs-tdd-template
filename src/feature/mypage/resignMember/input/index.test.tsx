import { toMock, toSpyWithMock } from "@/__testing__/helper";
import { loginRequiredPages } from "@/const/paths";
import { setCache } from "@/utils/cache";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as router from "next/router";
import IndexTemplate from "./index";

jest.mock("@/utils/cache");

describe("IndexTemplate", () => {
	const routerMock = jest.fn();

	afterEach(() => {
		jest.resetAllMocks();
		jest.restoreAllMocks();
	});

	function setup() {
		return {
			user: userEvent.setup(),
			view: render(<IndexTemplate />),
		};
	}

	describe("正常系", () => {
		describe("初期描画が変化していない", () => {
			const { view } = setup();
			expect(view.container).toMatchSnapshot();
		});

		describe("フォーム送信に成功する", () => {
			it("キャッシュに保存して、ページ遷移する", async () => {
				// Arrange
				const mockSetCache = toMock(setCache);
				const useRouterMockImplementation = () => ({
					push: routerMock,
				});
				toSpyWithMock(router, "useRouter", useRouterMockImplementation);

				// Act
				const { user, view } = setup();

				const reasonTypeSelectBox = await view.findByLabelText("退会理由");
				const reasonDetailInput = await view.findByLabelText("詳細");
				const agreeementCheckBox = await view.findByLabelText("同意する");
				await user.selectOptions(reasonTypeSelectBox, "その他");
				await user.type(reasonDetailInput, "使わないため");
				await user.click(agreeementCheckBox);

				const submitButton = await view.findByRole("button", {
					name: "退会する",
				});
				await user.click(submitButton);

				// Assert
				expect(mockSetCache).toHaveBeenCalledWith("resignMember", {
					reasonType: "OTHER",
					reasonDetail: "使わないため",
					agreement: true,
				});
				expect(routerMock).toHaveBeenCalledWith(
					loginRequiredPages.mypageResignMemberConfirm.path(),
				);
			});
		});
	});
});
