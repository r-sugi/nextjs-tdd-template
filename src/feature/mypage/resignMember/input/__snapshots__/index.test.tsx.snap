// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div>
    <form>
      <div>
        <label
          id="reasonType"
        >
          退会理由
        </label>
        <select
          aria-labelledby="reasonType"
          name="reasonType"
        >
          <option
            value=""
          >
            選択してください
          </option>
          <option
            value="NO_USE"
          >
            利用しないため
          </option>
          <option
            value="OTHER"
          >
            その他
          </option>
        </select>
      </div>
      <div>
        <label
          for="reasonDetail"
        >
          詳細
        </label>
        <textarea
          id="reasonDetail"
          name="reasonDetail"
        />
      </div>
      <div>
        <label
          for="agreement"
        >
          同意する
        </label>
        <input
          id="agreement"
          name="agreement"
          type="checkbox"
        />
      </div>
      <button
        disabled=""
        type="submit"
      >
        退会する
      </button>
    </form>
  </div>
</div>
`;
